### DISCUSSION TITLE
error in mcsd model fitting (DCPError)

### DISCUSSION THREAD

**Discussion by HilaGast** [dipy/dipy#2397] [Reconstruction] [answered]
## Description
Hi, I'm trying to fit mcsd model to my data and encounter the following error, any suggestions what might be the problem?
I succeed in running the model for other scans in the same protocol

```python
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\dipy\reconst\multi_voxel.py", line 33, in new_fit
    fit_array[ijk] = single_voxel_fit(self, data[ijk])
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\dipy\reconst\mcsd.py", line 310, in fit
    coeff = self.fitter(data)
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\dipy\reconst\mcsd.py", line 429, in __call__
    fodf_sh = solve_qp(self._P_mat, Q_mat, self._reg_mat, self._h_mat)
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\dipy\reconst\mcsd.py", line 393, in solve_qp
    prob.solve()
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\cvxpy\problems\problem.py", line 401, in solve
    There is no one choice of arguments which is perfect for every problem. If you are not
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\cvxpy\problems\problem.py", line 816, in _solve
    warm_start : bool, optional
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\cvxpy\problems\problem.py", line 505, in get_problem_data
    "l": cone_dims.nonpos,
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\cvxpy\problems\problem.py", line 702, in _construct_chain
    not candidates['qp_solvers']:
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\cvxpy\reductions\solvers\solving_chain.py", line 152, in construct_solving_chain
    reductions = _reductions_for_problem_class(problem, candidates, gp)
  File "C:\Users\HilaG\miniconda3\envs\LabPy\lib\site-packages\cvxpy\reductions\solvers\solving_chain.py", line 91, in _reductions_for_problem_class
    "Problem does not follow DCP rules. Specifically:\n" + append)
cvxpy.error.DCPError: Problem does not follow DCP rules. Specifically:
The objective is not DCP, even though each sub-expression is.
You are trying to minimize a function that is concave.
```

Windows-10-10.0.19041-SP0
Python 3.7.7

cvxpy           1.1.10
dipy            1.30
fury            0.6.0
h5py            2.10.0
nibabel         3.1.0
numpy           1.20.0
osqp            0.6.2.post0
scikit-learn    0.23.1
scipy           1.5.2
scs             2.1.2
vtk             9.0.1

Thank you all in advance

**Comment by skoudoro** [comment]
Hi @HilaGast,

I personally never encountered this kind of error. Maybe @ShreyasFadnavis @arokem @CHrlS98 know more about it.

Can you tell us more about your data and which preprocessing step are you applying before the modeling.

Thank you

**Comment by HilaGast** [comment]
> Hi @HilaGast,
> 
> I personally never encountered this kind of error. Maybe @ShreyasFadnavis @arokem @CHrlS98 know more about it.
> 
> Can you tell us more about your data and which preprocessing step are you applying before the modeling.
> 
> Thank you

Hi, thanks for answering.

My data scan protocol consists of multi-shelled (b: 0, 1000, 2000 & 4000) HARDI (186 directions) dMRI protocol (Big Delta: 45 , Small Delta: 15).

For analysis I'm using data after eddy-current and movement correction (FSL based - eddy) and tissue classification (using FAST in FSL) for response function (response_from_mask_msmt and than multy_shell_fiber_response).

It is noteworthy that I successfully analyzed different scans using the same protocol before, and that the problematic scan was analyzed using CSD without errors.

Attempting to treat DCPError similar to SolverError didn't seem to work since the problem seems to be with a large number of voxels.

Hila

**Comment by arokem** [comment]
Hi Hila - thanks for reporting this. Could you please share the values of
your bvals and the values of the signal in one of the voxels where this is
happening?

**Comment by HilaGast** [comment, marked as answer]
> Hi Hila - thanks for reporting this. Could you please share the values of your bvals and the values of the signal in one of the voxels where this is happening?
> […](#)
> On Tue, May 4, 2021 at 12:52 AM HilaGast ***@***.***> wrote: Hi @HilaGast <https://github.com/HilaGast>, I personally never encountered this kind of error. Maybe @ShreyasFadnavis <https://github.com/ShreyasFadnavis> @arokem <https://github.com/arokem> @CHrlS98 <https://github.com/CHrlS98> know more about it. Can you tell us more about your data and which preprocessing step are you applying before the modeling. Thank you Hi, thanks for answering. My data scan protocol consists of multi-shelled (b: 0, 1000, 2000 & 4000) HARDI (186 directions) dMRI protocol (Big Delta: 45 , Small Delta: 15). For analysis I'm using data after eddy-current and movement correction (FSL based - eddy) and tissue classification (using FAST in FSL) for response function (response_from_mask_msmt and than multy_shell_fiber_response). It is noteworthy that I successfully analyzed different scans using the same protocol before, and that the problematic scan was analyzed using CSD without errors. Attempting to treat DCPError similar to SolverError didn't seem to work since the problem seems to be with a large number of voxels. Hila — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#2383 (comment)](https://github.com/dipy/dipy/issues/2383#issuecomment-831750445)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAA46NXPDLQX5VOH4F347LTTL6RUVANCNFSM436N7ZMQ> .

It turns out that in the pre-processing, a few voxels received negative values, most of them outside the brain. Their reset solved the problem.

Thanks!
Hila
